ipmininet/cli.py

```diff
@@ -31,3 +33,60 @@ def do_ips(self, line):
             finally:
                 lg.info(n, '|', l)
         lg.info('\n')
+
+    def default(self, line):
+        """Called on an input line when the command prefix is not recognized.
+        Overridden to run shell commands when a node is the first CLI argument.
+        Past the first CLI argument, node names are automatically replaced with
+        corresponding addresses if possible.
+        We select only one IP version for these automatic replacements. The IP version
+        chosen is first restricted by the addresses available on the first node.
+        Then, we choose the IP version that enables as many replacements as possible.
+        We use IPv4 as a tie-break."""
+
+        first, args, line = self.parseline(line)
+
+        if first in self.mn:
+            if not args:
+                print("*** Enter a command for node: %s <cmd>" % first)
+                return
+            node = self.mn[first]
+            rest = args.split(' ')
+
+            # Identify which IP protocols can be used by the node on which the command is issued
+            first_v4 = False
+            first_v6 = False
+            for itf in realIntfList(node):
+                if itf.updateIP() is not None:
+                    first_v4 = True
+                itf.updateIP6()
+                for _ in itf.ip6s(exclude_lls=True):
+                    first_v6 = True
+                    break
+
+            # Identify the possible substitutions
+            ipv4_map = {}
+            ipv6_map = {}
+            for r in rest:
+                if r in self.mn:
+                    other_node = self.mn[r]
+                    for itf in realIntfList(other_node):
+                        if first_v4 and itf.updateIP() is not None:
+                            ipv4_map[r] = itf.ip
+                        if first_v6:
+                            itf.updateIP6()
+                            for ip6 in itf.ip6s(exclude_lls=True):
+                                ipv6_map[r] = str(ip6.ip)
+                                break
+
+            ip_map = ipv4_map if len(ipv4_map) >= len(ipv6_map) else ipv6_map
```

Comment:
This could be rewritten better to signal an error for dual stacked networks when there is a conflict with the available ip version:

(starting at line 56//Not tested)
```python
def address_pair(n, use_v4=True, use_v6=True):
    v4 = v6 = None
    for itf in realIntfList(n):
        if use_v4 and v4 is None:
            v4 = itf.updateIP()
        if use_v6 and v6 is None:
            itf.updateIP6()
            v6 = next(itf.ip6s(exclude_lls=True), None)
        if (not use_v4 or v4 is not None) and (not use_v6 or v6 is not None):
            break
    return v4, v6
 
            
hops = [h for h in rests if h in self.mn]
v4_support, v6_support = address_pair(first)
v4_map = {}
v6_map = {}
for hop in hops:
    v4_map[hop], v6_map[hop] = address_pair(hop, v4_support is not None, v6_support is not None)
ip_map = v4_map if len(v4_map) >= len(v6_map) else v6_map
if len(ip_map) < len(hops):
    missing = filter(lambda h: h not in ip_map, hops)
    version = 'IPv4' if v4_support else 'IPv6'
    lg.error('*** Nodes', missing, 'have no', version, 'address! Cannot execute the command.\n')
    return
node.sendCmd(' '.join(ip_map[h] for h in hops))
self.waitForNode(node)
```